---
title: "Workshop Setup Instructions"
author: "Adele B and Paul H"
date: "2022-11-08"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Get the workshop material and data

In RStudio **create a new project**. This ensures all the files for this workshop are placed in their own folder.

Once you've created a new project, run the following R code to download the workshop material and dataset:

```{r eval=FALSE}
## Download files we will be using
options(timeout=3600)
filenames <- c("pbmc3k_tutorial.R", "kang2018.rds")
url <- "https://raw.githubusercontent.com/MonashBioinformaticsPlatform/PBMC-Single-Cell-Workshop-2022/main/vignettes/" 
for(filename in filenames)
    download.file(paste0(url,filename), filename)

## Download and untar the raw data
download.file(
    "http://10x.files.s3-us-west-2.amazonaws.com/samples/cell/pbmc3k/pbmc3k_filtered_gene_bc_matrices.tar.gz",
    "pbmc3k_filtered_gene_bc_matrices.tar.gz")

untar("pbmc3k_filtered_gene_bc_matrices.tar.gz")
```

## Package Installation
 
For this workshop, several packages need to be installed.

BiocManager likes to update installed packages, but we have disabled this in the R code below. If your installation fails, then you might need to turn updates on. Note that if there are a large number of packages that BiocManager wants to update it can take several hours.

```{r, eval=FALSE}
## Install required packages for Seurat:
install.packages(c("Seurat", "dplyr", "remotes", "R.utils"))

# 2022-10-30 Harmony failed checks on CRAN. Use dev version for now.
remotes::install_github("immunogenomics/harmony")

## Install required Bioconductor packages
if (!requireNamespace("BiocManager", quietly = TRUE))
  install.packages("BiocManager")

BiocManager::install(c('SingleR', 'celldex',
                       'BiocGenerics', 'DelayedArray', 'DelayedMatrixStats',
                       'limma', 'S4Vectors', 'SingleCellExperiment',
                       'SummarizedExperiment', 'batchelor', 'Matrix.utils'),
                     update=FALSE)

## Install dev versions of some interactive graphics packages
remotes::install_github("plotly/plotly.R")
remotes::install_github("pfh/langevitour")
```


Troubleshooting installation on M1 Mac with R.4.2:

1. Do not install any packages from source - select `no` when asked.

2. Additional packages to install to load Seurat & Harmony:

```{r, eval=FALSE}
install.packages(c("rgeos", "spatstat.core", "RcppEigen", "RcppArmadillo"))
```

3. Harmony will not install and returns an error that looks like something like this:

```
ld: warning: directory not found for option '-L/opt/R/arm64/gfortran/lib/gcc/aarch64-apple-darwin20.6.0/12.0.1'
ld: library not found for -lemutls_w
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make: *** [harmony.so] Error 1
ERROR: compilation failed for package ‘harmony’
* removing ‘/Library/Frameworks/R.framework/Versions/4.2-arm64/Resources/library/harmony’
Warning message:
In i.p(...) :
  installation of package ‘/var/folders/1k/bbv_y78s4zj0m29mb6fwj3km69dgdb/T//Rtmp9jOOMQ/file83ee3a886f7c/harmony_0.1.1.tar.gz’ had non-zero exit status
```

Solution: install [GNU Fortran compiler](https://mac.r-project.org/tools/). Following the instructions for that link in regards to Apple Silicon Macs (M1 and above):

i. Download the experimental branch of GNU Fortran (run the code below in your terminal)

```{bash, eval=FALSE}
wget https://mac.r-project.org/tools/gfortran-12.0.1-20220312-is-darwin20-arm64.tar.xz
```

ii. This binary unpacks into /opt/R/arm64/gfortran plus a symbolic link in /opt/R/arm64/bin/gfortran. To unpack use:
```{bash, eval=FALSE}
tar fxz gfortran-12.0.1-20220312-is-darwin20-arm64.tar.xz -C /
```
If this fails because you do not have permission, run it with sudo as so:

```{bash, eval=FALSE}
sudo tar fxz gfortran-12.0.1-20220312-is-darwin20-arm64.tar.xz -C /
```
It will ask for a password - this should be the same password as your user account. 

iii. Install harmony with devtools:

```{r, eval=FALSE}
install.packages("devtools")
devtools::install_github("immunogenomics/harmony")
```